### ISSUE TITLE
how should dynamic dispatch interact with tertiary methods?

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#16854]
This issue is a spin-off from https://github.com/chapel-lang/chapel/issues/16852#issuecomment-742875588

Other elements of our current direction for method visibility have tertiary methods only available when they are visible (e.g. from `use ModuleDefiningTertiaryMethod` possibly with `only SomeType`). However, for methods that are dynamically dispatched, should the visibility of an `override proc` affect the dynamic dispatch?

Here is an example to think about:

``` chapel
module DefinesParentChild {
  class Parent {
    proc method() {
      writeln("Parent.method()");
    }
  }
  class Child: Parent { }
}

module DefinesOverride {
  use DefinesParentChild;

  override proc Child.method() {
    writeln("Child.method()");
  }
}

module Tests {
  { // case 1
    use DefinesParentChild;
    var child: owned Child = new Child();
    child.method(); // does it run Parent.method or Child.method ?
  }
  { // case 2
    use DefinesParentChild;
    var child: owned Parent = new Child();
    child.method(); // does it run Parent.method or Child.method ?
  }
  { // case 3
    use DefinesParentChild;
    use DefinesOverride;
    var child: owned Child = new Child();
    child.method(); // does it run Parent.method or Child.method ?
  }
  { // case 4
    use DefinesParentChild;
    use DefinesOverride;
    var child: owned Parent = new Child();
    child.method(); // does it run Parent.method or Child.method ?
  }
}
```

What does it do today?
<details>
* case 1 runs Child.method
* case 2 runs Child.method
* case 3 results in an overload sets error
* case 4 runs Child.method
</details>

Possible directions than I can think of:
 * it is an error to define a tertiary `override proc`
 * such an `override proc` is always called if the dynamic type is compatible
 * the dynamic dispatch uses information about visibility in the call at runtime to more closely match the non-dynamic behavior

**Comment by mppf** [comment]
My thoughts about what to do:

* It seems reasonable to make it an error if testing does not indicate common patterns that will disallow.
* Maybe always calling the `override proc` is reasonable. Is there some analogy to the way that something with compile-time type `Parent` can invoke methods on the dynamic type (e.g. `Child`) even if neither `Child` nor these methods are visible? But, maybe this is not desired for software engineering reasons - an author of a `Child` type that does not define all overrideable methods can have code working, but then if a 3rd module is added that defines them, the other calls (with no relation to the 3rd module) now change behavior - this is arguably a kind of "hijacking".
* I really don't want dynamic dispatch to depend on visibility.

**Comment by lydia-duncan** [comment]
I lean towards making it an error to define a tertiary `override proc`.  I think allowing it would lead to subtle bugs.

The case you have in the original post has the benefit of all the code being in the same file.  The following code fails to compile today due to overload sets, but potentially could be tricked into compiling in the future with separate compilation - in that case, it would probably have different behavior for the two usesChild functions.  Should that be okay?

File 1
```chapel
module DefinesParentChild {
  class Parent {
    proc method() {
      writeln("Parent.method()");
    }
  }
  class Child: Parent { }
}

module SomeLib1 {
  use DefinesParentChild;
  
  proc usesChild() {
    var child: owned Child = new Child();
    child.method();
  }
}
```

File 2
```chapel
module DefinesOverride {
  use DefinesParentChild;

  override proc Child.method() {
    writeln("Child.method()");
  }
}
module SomeLib2 {
  use DefinesParentChild;
  use DefinesOverride;
  
  proc usesChild() {
    var child: owned Child = new Child();
    child.method();
  }
}
```

File 3
```chapel
module Main {
  import SomeLib1, SomeLib2;

  proc main() {
     SomeLib1.usesChild();
     SomeLib2.usesChild();
  }
}
```

**Comment by bradcray** [comment]
> It seems reasonable to make it an error if testing does not indicate common patterns that will disallow.

I agree with this.  And if, down the road, people come up with compelling cases that need such patterns and we can make them work logically, we can reconsider then.

**Comment by bradcray** [comment]
(BTW, in addition to not permitting tertiary `override` methods, I was thinking we'd also need to make it illegal to define parent-class methods that shared the same name as a non-`override` method on the child class?  That is:

```chapel
module M {
  class Parent {
  }

  class Child: Parent {
    proc foo() { writeln("In Child.foo()"); }
  }
}

module Extend {
  use M;
  proc Parent.foo() { writeln("In Parent.foo()"); }  // This should be illegal
}

module Main {
  proc main() {
    use M, Extend;

    var myP = new Parent();
    var myC: Parent = new Child();
    var myC2 = new Child();

    myP.foo();                                                                
    myC.foo();                                                                
    myC2.foo();
  }
}
```

Happily, this seems to already be prevented today.

**Comment by cassella** [comment]
Further arguing against allowing tertiary `override` methods is the spectre of different tertiary overrides of the same method defined in different modules.  If I add a new module to the original example (with case 3 removed):

```chapel
module DefinesOtherOverride {
  use DefinesParentChild;

  override proc Child.method() {
    writeln("other Child.method()");
  }
}
```

1. It has no effect if it's not used -- the program compiles and all three existing cases report `Child.method()`.

2. With a new case 5 (a copy of case 4, but with a `use` of `DefinesOtherOverride` instead of `DefinesOverride`) all four cases report `Child.method()`, even case 5.

As long as the definition of `DefinesOtherOverride` comes after that of `DefinesOverride`.  (The order of case 5 vs the other cases doesn't have an effect.)

3. Swapping the order of the definitions changes the behavior -- with `DefinesOtherOverride` defined first, all four cases report `other Child.method()`.  Again, the order of the cases has no effect.

4. If I remove cases 4 and 5 (and thus all `use`s of the overriding modules), cases 1 and 2 report `Parent.method()`.


Apart from the current behavior, what should the behavior be, if two scopes can create a `Child` object with a different override visible, and then make those objects visible to a third scope with no override visible, should those two objects of ostensibly the same type result in different methods being called?

Should an object created with no override visible have the method overridden?

**Comment by mppf** [comment]
Based on experience in writing some compiler improvements, I actually really like the idea of being able to have dynamically dispatched tertiary methods because it is an extremely easy way to implement the visitor pattern. In that case, the tertiary method is overriding another tertiary method. So, I don't think it should always be an error, but rather it should be an error if a tertiary method is overriding a primary/secondary method.

**Comment by lydia-duncan** [comment]
Should we further limit it to "overriding a tertiary method *in the same module*"?  Or would that eliminate much of your intended use case?

**Comment by bradcray** [comment]
> Should we further limit it to "overriding a tertiary method in the same module"?

That seems appropriate to me.  It seems like any sort of cross-module dynamic dispatch scenarios are similar regardless of whether the method being overridden is primary, secondary, or tertiary.

**Comment by mppf** [comment]
> Should we further limit it to "overriding a tertiary method in the same module"? Or would that eliminate much of your intended use case?

I'm sorry but I'm not able to make any sense of this. Same module as what? Can you show an example that demonstrates the difference between the proposals?

My intended use case is below and it creates a tertiary method and a tertiary override.

``` chapel
module Library {
  class C {
    proc method() { }
  }
  class D: C { }
  class E: C { }
  var myD = new D();
  myD.method();
}
module OKOther {
  use Library;
  // these methods are not defined on C/D in Library
  // and so they can't affect code in there
  proc C.otherOperation() { writeln("in C.otherOperation"); }
  override proc D.otherOperation() { writeln("in D.otherOperation"); }

  proc main() {
    var d: C = new D();
    var e: C = new E();
    d.otherOperation();
    e.otherOperation();
  }
}
```

**Comment by lydia-duncan** [comment]
> I'm sorry but I'm not able to make any sense of this. Same module as what?

I'm saying the tertiary method should be defined in the same module as the tertiary method it is overriding.  Your example fulfills that condition as `C.otherOperation` and `D.otherOperation` are defined in the same module.  The following example would not:

```chapel
module Library {
  class C {
    proc method() { }
  }
  class D: C { }
  class E: C { }
  var myD = new D();
  myD.method();
}
module DefinesTertOnC {
  use Library;

  // this method is not defined on C in Library
  // and so it can't affect code in there
  proc C.otherOperation() { writeln("in C.otherOperation"); }
}
module DefinesTertOnD {
  use Library;
  use DefinesTertOnC;

  // this method is not defined on C in Library
  // and so it can't affect code in there
  override proc D.otherOperation() { writeln("in D.otherOperation"); }
}
```
Code that uses DefinesTertOnC would have different behavior than code that uses DefinesTertOnD, which brings us back to the original problem.
